wannierberri/__Data_K.py

```diff
@@ -28,59 +28,87 @@ def _rotate_matrix(X):
         self.use_wcc_phase=system.use_wcc_phase
-        ## TODO : create the plans externally, one per process 
-#        print( "iRvec in data_K is :\n",self.iRvec)
-        #self.fft_R_to_k=FFT_R_to_k(self.iRvec,self.NKFFT,self.num_wann,self.wannier_centres,numthreads=npar if npar>0 else 1,lib=fftlib,convention=system.convention)
-        self.fft_R_to_k=FFT_R_to_k(self.iRvec,self.NKFFT,self.num_wann,self.wannier_centres_reduced,self.real_lattice,
+        self.fft_R_to_k=FFT_R_to_k(self.system.iRvec,self.NKFFT,self.system.num_wann,
+               self.system.wannier_centres_reduced,self.system.real_lattice,
                 numthreads=npar if npar>0 else 1,lib=fftlib,use_wcc_phase=self.use_wcc_phase)
         self.Emin=system.Emin
         self.Emax=system.Emax
         self.poolmap=pool(npar_k)[0]
 
         
         if self.use_wcc_phase:
-            w_centres_diff = np.array([[j-i for j in self.wannier_centres_reduced] for i in self.wannier_centres_reduced])
-            expdK=np.exp(2j*np.pi*(system.iRvec[None,None,:,:] +w_centres_diff[:,:,None,:]).dot(dK))
+            w_centres_diff = np.array([[j-i for j in self.system.wannier_centres_reduced] for i in self.system.wannier_centres_reduced])
+            self.expdK=np.exp(2j*np.pi*(self.system.iRvec[None,None,:,:] +w_centres_diff[:,:,None,:]).dot(dK))
         else:
-            expdK=np.exp(2j*np.pi*system.iRvec.dot(dK))[None,None,:]
-        self.HH_R=system.HH_R*expdK
+            self.expdK=np.exp(2j*np.pi*self.system.iRvec.dot(dK))[None,None,:]
+        self.HH_R=system.HH_R*self.expdK
         self.dK=dK
+
  
-        
-        for X in ['AA','BB','CC','SS','SA','SHA','SR','SH','SHR']:
-            XR=X+'_R'
-            hasXR='has_'+X+'_R'
-            vars(self)[XR]=None
-            vars(self)[hasXR]=False
-            if XR in vars(system):
-              if vars(system)[XR] is not  None:
-                if X in ['SA','SHA','SR','SHR']:
-                    vars(self)[XR]=vars(system)[XR]*expdK[:,:,:,None,None]
-                else:
-                    vars(self)[XR]=vars(system)[XR]*expdK[:,:,:,None]
-                vars(self)[hasXR]=True
 
 
+###########################################
+###   Now the **_R objects are evaluated only on demand 
+### - as Lazy_property (if used more than once) 
+###   as property   - iif used only once
+###   let's write them explicitly, for better code readability
+###########################
+
+    @lazy_property.LazyProperty
+    def AA_R(self):
+        return self.system.AA_R*self.expdK[:,:,:,None]
+
+    @lazy_property.LazyProperty
+    def BB_R(self):
+        return self.system.BB_R*self.expdK[:,:,:,None]
+
+    @lazy_property.LazyProperty
+    def CC_R(self):
+        return self.system.CC_R*self.expdK[:,:,:,None]
+
+    @lazy_property.LazyProperty
+    def SS_R(self):
+        return self.system.SS_R*self.expdK[:,:,:,None]
+
+    @property
+    def SH_R(self):
+        return self.system.SH_R*self.expdK[:,:,:,None]
+
```

Comment:
Why is SH_R (and all other spin Hall related matrices) a `@property` while others are `@LazyProperty`?

Reply:
because they are used only once in the class, and the use is in another lazy_property `SA_H`, `SHA_H` or `shc_B_H` so, there is no need to store `SH_R`. Also removed corresponding  copy() for that reason.  `AA_R` , `BB_R`, `CC_R` are used at least twice.